lib/config/config_test.go

```diff
@@ -0,0 +1,191 @@
+package config
+
+/*
+ The tests within this file provide minimal testing for Robotest's test configuration DSL,
+ and also serve as examples of how custom test parameters can be declared, validated, and
+ and have defaults provided.
+*/
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/gravitational/robotest/infra/gravity"
+	"github.com/gravitational/trace"
+)
+
+// defaultTimeout is meaningless test data, it could be any duration
+var defaultTimeout = 5 * time.Second
+
+// testParam is an artificial struct similar to suite/sanity/install.go:installParam.
+type testParam struct {
+	UID       uint             `json:"uid"`
+	Username  string           `json:"user"`
+	Operation *nestedTestParam `json:"operation"`
+}
+
+// CheckAndSetDefaults provides some basic default logic validation for testParam.
+//
+// This is provided as an example of what sort of custom validation can be done
+// by fulfilling the Validator interface.
```

Comment:
This is a wierd one.  Go recommends `*er` for single method interfaces.  For example:

`Reader` -> `Read()`

Thus I'd expect 

`validator` -> `Validate()` and `defaulter` -> `Default()`

Previously, this function only validated data, which meant it could accept a `struct` value in the receiver.  However, now that it may set default values in the struct, it must have a receiver of `*struct` such that the changed defaults can escape the method call.

I felt it was misleading to have a method called `Validate()` that would alter the receiver -- to me validate should never change things, only return an error if it is invalid.  So I went with your suggestion of `CheckAndSetDefaults` as the method name -- a gravitational-ism that seems reasonable.  However the corresponding interface name `checkAndSetDefaultser` felt not great... which is how I arrived to `defaulter`.  At least it shared a word with the method name.

Aside: 

I even explored having two seperate interfaces which test parameters could implement:

```go
type validator interface {
    Validate() // receiver is expected to be interface {} (struct)
}

type defaulter interface {
    setDefaults() // receiver is expected to be interface {} (*struct)
}
```

But this seemed pretty overarchitechted to me.

Reply:
Oh, I remember why I chose `defaulter` now: the `validator` name is taken by https://github.com/go-playground/validator in  `lib/config/config.go` as seen on https://github.com/gravitational/robotest/pull/235/files/29f0bd3beca2a490084ae94ef1551935dae64e08#diff-319b5feabc77b066118cda94584168a8R137

I suppose I could export it as `Validator`?  Since code in `suite/sanity/*` is being written to depend on `CheckAndSetDefaults` it probably should be part of the public API.

Reply:
Comment updated to match code in https://github.com/gravitational/robotest/pull/235#discussion_r434014167.  If you prefer the exported `Validator`, let me know and I'll make the change.